language-plutus-core/src/Language/PlutusCore/Rename/Monad.hs

```diff
@@ -0,0 +1,135 @@
+
+import           PlutusPrelude
+
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Quote
+
+import           Control.Lens
+import           Control.Monad.Reader
+
+-- | The monad the renamer runs in.
+newtype RenameT ren m a = RenameT
+    { unRenameT :: ReaderT ren m a
+    } deriving
+        ( Functor, Applicative, Alternative, Monad
+        , MonadReader ren
+        , MonadQuote
+        )
+
+-- | A renaming is a mapping from old uniques to new ones.
+newtype Renaming unique = Renaming
+    { unRenaming :: UniqueMap unique unique
+    } deriving (Semigroup, Monoid)
+
+-- | A class that specifies which 'Renaming' a @ren@ has inside.
+-- A @ren@ can contain several 'Renaming's (like 'Scoped', for example).
+class Coercible unique Unique => HasRenaming ren unique where
+    renaming :: Lens' ren (Renaming unique)
+
+-- | Scoping-aware mapping from locally unique uniques to globally unique uniques.
+data ScopedRenaming = ScopedRenaming
+    { _scopedRenamingTypes :: Renaming TypeUnique
+    , _scopedRenamingTerms :: Renaming TermUnique
+    }
+
+makeLenses ''ScopedRenaming
+
+type ScopedRenameT = RenameT ScopedRenaming
+
+instance Semigroup ScopedRenaming where
+    ScopedRenaming types1 terms1 <> ScopedRenaming types2 terms2 =
+        ScopedRenaming (types1 <> types2) (terms1 <> terms2)
+
+instance Monoid ScopedRenaming where
+    mempty = ScopedRenaming mempty mempty
+
+instance (Coercible unique1 Unique, unique1 ~ unique2) =>
+        HasRenaming (Renaming unique1) unique2 where
+    renaming = id
+
+instance HasRenaming ScopedRenaming TypeUnique where
+    renaming = scopedRenamingTypes . renaming
+
+instance HasRenaming ScopedRenaming TermUnique where
+    renaming = scopedRenamingTerms . renaming
+
+-- | Run a 'RenameT' computation with a supplied @ren@.
+runRenameT :: ren -> RenameT ren m a -> m a
+runRenameT ren (RenameT a) = runReaderT a ren
+
+-- | Run a 'RenameT' computation with a 'mempty' renaming.
+mrunRenameT :: Monoid ren => RenameT ren m a -> m a
+mrunRenameT = runRenameT mempty
+
+-- | Run a 'RenameT' computation with an empty 'Renaming'.
+-- A specialized version of 'mrunRenameT' needed for the sake of type inference.
+runDirectRenameT :: RenameT (Renaming unique) m a -> m a
+runDirectRenameT = mrunRenameT
+
+-- | Map the underlying representation of 'Renaming'.
+mapRenaming
+    :: (UniqueMap unique unique -> UniqueMap unique unique)
+    -> Renaming unique
+    -> Renaming unique
+mapRenaming = coerce
+
+-- | Save the mapping from the @unique@ of a name to a new @unique@.
+insertByNameM
+    :: (HasUnique name unique, HasRenaming ren unique)
+    => name -> unique -> ren -> ren
+insertByNameM name = over renaming . mapRenaming . insertByName name
+
+-- | Look up the new unique a name got mapped to.
+lookupNameM
+    :: (HasUnique name unique, HasRenaming ren unique, Monad m)
+    => name -> RenameT ren m (Maybe unique)
+lookupNameM name = asks $ lookupName name . unRenaming . view renaming
```

Comment:
I've been enjoying some more `lens` features recently. You could define instances of `Indexed` for `UniqMap` and `Renaming`, and then write something like `use (renaming . at name)`.

(That said, I'm not sure whether I think we should be using more lens stuff or not.)

Reply:
I'll try it out.

Reply:
> You could define instances of Indexed for `UniqMap` and `Renaming`, and then write something like `use (renaming . at name)`.

Nope. We insert and lookup into `UniqMap` by both uniques and names and the `At` type class does not support multiple indices. We could only support uniques and extract uniques from names using the `unique` lens, but I'm not sure what this additional complexity buys us.

For `Renaming` we use only those functions that deal with names, but we can't define appropriate instances of `Index`, `At`, etc, because even though functions deal with names, the underlying representation only knows about uniques and so we can't define something like

```haskell
type instance Index (Renaming unique) = name
```

because what would `name` be?

(BTW, we're in `Reader` rather than `State`, so it's `view` rather than `use`)

Reply:
So I'd prefer to keep some old-fashioned Haskell here.

Reply:
Reasonable.